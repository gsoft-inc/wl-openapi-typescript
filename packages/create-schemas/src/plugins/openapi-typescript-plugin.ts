import openapiTS, { astToString, type OpenAPI3 } from "openapi-typescript";
import type { Plugin } from "./plugin.ts";

export const openapiTypeScriptId = "internal:openapi-typescript-plugin";

export const openapiTypeScriptFilename = "types.ts";

export function openapiTypeScriptPlugin(): Plugin {
    return {
        name: "internal:openapi-typescript-plugin",
        async buildStart({ config, emitFile, document }) {
            const ast = await openapiTS(document as OpenAPI3, {
                silent: true,
                ...config.openApiTsOptions
            });

            emitFile({
                id: openapiTypeScriptId,
                filename: openapiTypeScriptFilename,
                code: astToString(ast)
            });
        }
    };
}
