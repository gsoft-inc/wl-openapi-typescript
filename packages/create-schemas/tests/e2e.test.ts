import {
    createTemporaryFolder,
    dataFolder,
    runCompiledBin
} from "./fixtures.ts";
import { assert, describe, test } from "vitest";
import { join } from "node:path";
import { pathToFileURL } from "node:url";
import { openapiTypeScriptFilename } from "../src/plugins/openapi-typescript-plugin.ts";

const timeout = 30 * 1000; // 30 seconds

describe.concurrent("e2e", () => {
    test(
        "officevice.yaml / file paths",
        async ({ expect, onTestFinished }) => {
            const tempFolder = await createTemporaryFolder({ onTestFinished });

            const result = await runCompiledBin({
                source: join(dataFolder, "officevice.yaml"),
                outdir: tempFolder
            });

            const typesFile = result.find(file => file.filename === openapiTypeScriptFilename);
            assert(typesFile);
            expect(typesFile.code).toMatchSnapshot(); 
        },
        timeout
    );

    test(
        "officevice.yaml / file URLs",
        async ({ expect, onTestFinished }) => {
            const tempFolder = await createTemporaryFolder({ onTestFinished });

            const result = await runCompiledBin({
                source: pathToFileURL(join(dataFolder, "officevice.yaml")).toString(),
                outdir: tempFolder
            });

            const typesFile = result.find(file => file.filename === openapiTypeScriptFilename);
            assert(typesFile);
            expect(typesFile.code).toMatchSnapshot(); 
        },
        timeout
    );

    test(
        "officevice.yaml / relative path",
        async ({ expect, onTestFinished }) => {
            const tempFolder = await createTemporaryFolder({ onTestFinished });

            const result = await runCompiledBin({
                cwd: dataFolder,
                source: "officevice.yaml",
                outdir: tempFolder
            });

            const typesFile = result.find(file => file.filename === openapiTypeScriptFilename);
            assert(typesFile);
            expect(typesFile.code).toMatchSnapshot(); 
        },
        timeout
    );

    test(
        "petstore.json",
        async ({ expect, onTestFinished }) => {
            const tempFolder = await createTemporaryFolder({ onTestFinished });

            const result = await runCompiledBin({
                source: join(dataFolder, "petstore.json"),
                outdir: tempFolder
            });

            const typesFile = result.find(file => file.filename === openapiTypeScriptFilename);
            assert(typesFile);
            expect(typesFile.code).toMatchSnapshot(); 
        },
        timeout
    );

    test(
        "petstore.json / remote URL",
        async ({ expect, onTestFinished }) => {
            const tempFolder = await createTemporaryFolder({ onTestFinished });

            const result = await runCompiledBin({
                source: "https://petstore3.swagger.io/api/v3/openapi.json", 
                outdir: tempFolder
            });

            const typesFile = result.find(file => file.filename === openapiTypeScriptFilename);
            assert(typesFile);
            expect(typesFile.code).toMatchSnapshot();
        },
        timeout
    );
});
