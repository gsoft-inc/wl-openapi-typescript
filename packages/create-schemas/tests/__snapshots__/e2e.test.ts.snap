// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`e2e > officevibe local yaml 1`] = `
"export interface paths {
    "/good-vibes-points/{userId}": {
        parameters: {
            query?: never;
            header?: never;
            path?: never;
            cookie?: never;
        };
        /** Get the current number of good vibe for a user */
        get: operations["GetGoodVibesPoint"];
        put?: never;
        post?: never;
        delete?: never;
        options?: never;
        head?: never;
        patch?: never;
        trace?: never;
    };
}
export type webhooks = Record<string, never>;
export interface components {
    schemas: {
        GetGoodVibePointsResult: {
            /** Format: int32 */
            point: number;
        };
        ProblemDetails: {
            type?: string | null;
            title?: string | null;
            /** Format: int32 */
            status?: number | null;
            detail?: string | null;
            instance?: string | null;
            [key: string]: unknown;
        };
    };
    responses: never;
    parameters: never;
    requestBodies: never;
    headers: never;
    pathItems: never;
}
export type $defs = Record<string, never>;
export interface operations {
    GetGoodVibesPoint: {
        parameters: {
            query?: never;
            header?: never;
            path: {
                userId: string;
            };
            cookie?: never;
        };
        requestBody?: never;
        responses: {
            /** @description Success */
            200: {
                headers: {
                    [name: string]: unknown;
                };
                content: {
                    "application/json": components["schemas"]["GetGoodVibePointsResult"];
                };
            };
            /** @description Bad Request */
            400: {
                headers: {
                    [name: string]: unknown;
                };
                content: {
                    "application/json": components["schemas"]["ProblemDetails"];
                };
            };
        };
    };
}
export type GetGoodVibePointsResult = components["schemas"]["GetGoodVibePointsResult"];
export type ProblemDetails = components["schemas"]["ProblemDetails"];

export type Endpoints = keyof paths;
"
`;

exports[`e2e > officevibe.yaml 1`] = `
"export interface paths {
    "/good-vibes-points/{userId}": {
        parameters: {
            query?: never;
            header?: never;
            path?: never;
            cookie?: never;
        };
        /** Get the current number of good vibe for a user */
        get: operations["GetGoodVibesPoint"];
        put?: never;
        post?: never;
        delete?: never;
        options?: never;
        head?: never;
        patch?: never;
        trace?: never;
    };
}
export type webhooks = Record<string, never>;
export interface components {
    schemas: {
        GetGoodVibePointsResult: {
            /** Format: int32 */
            point: number;
        };
        ProblemDetails: {
            type?: string | null;
            title?: string | null;
            /** Format: int32 */
            status?: number | null;
            detail?: string | null;
            instance?: string | null;
            [key: string]: unknown;
        };
    };
    responses: never;
    parameters: never;
    requestBodies: never;
    headers: never;
    pathItems: never;
}
export type $defs = Record<string, never>;
export interface operations {
    GetGoodVibesPoint: {
        parameters: {
            query?: never;
            header?: never;
            path: {
                userId: string;
            };
            cookie?: never;
        };
        requestBody?: never;
        responses: {
            /** @description Success */
            200: {
                headers: {
                    [name: string]: unknown;
                };
                content: {
                    "application/json": components["schemas"]["GetGoodVibePointsResult"];
                };
            };
            /** @description Bad Request */
            400: {
                headers: {
                    [name: string]: unknown;
                };
                content: {
                    "application/json": components["schemas"]["ProblemDetails"];
                };
            };
        };
    };
}
export type GetGoodVibePointsResult = components["schemas"]["GetGoodVibePointsResult"];
export type ProblemDetails = components["schemas"]["ProblemDetails"];

export type Endpoints = keyof paths;
"
`;

exports[`e2e > petstore local json 1`] = `
"export interface paths {
    "/pets": {
        parameters: {
            query?: never;
            header?: never;
            path?: never;
            cookie?: never;
        };
        /** List all pets */
        get: operations["listPets"];
        put?: never;
        /** Create a pet */
        post: operations["createPets"];
        delete?: never;
        options?: never;
        head?: never;
        patch?: never;
        trace?: never;
    };
    "/pets/{petId}": {
        parameters: {
            query?: never;
            header?: never;
            path?: never;
            cookie?: never;
        };
        /** Info for a specific pet */
        get: operations["showPetById"];
        put?: never;
        post?: never;
        delete?: never;
        options?: never;
        head?: never;
        patch?: never;
        trace?: never;
    };
}
export type webhooks = Record<string, never>;
export interface components {
    schemas: {
        Pet: {
            /** Format: int64 */
            id: number;
            name: string;
            tag?: string;
        };
        Pets: components["schemas"]["Pet"][];
        Error: {
            /** Format: int32 */
            code: number;
            message: string;
        };
    };
    responses: never;
    parameters: never;
    requestBodies: never;
    headers: never;
    pathItems: never;
}
export type $defs = Record<string, never>;
export interface operations {
    listPets: {
        parameters: {
            query?: {
                /** @description How many items to return at one time (max 100) */
                limit?: number;
            };
            header?: never;
            path?: never;
            cookie?: never;
        };
        requestBody?: never;
        responses: {
            /** @description A paged array of pets */
            200: {
                headers: {
                    /** @description A link to the next page of responses */
                    "x-next"?: string;
                    [name: string]: unknown;
                };
                content: {
                    "application/json": components["schemas"]["Pets"];
                };
            };
            /** @description unexpected error */
            default: {
                headers: {
                    [name: string]: unknown;
                };
                content: {
                    "application/json": components["schemas"]["Error"];
                };
            };
        };
    };
    createPets: {
        parameters: {
            query?: never;
            header?: never;
            path?: never;
            cookie?: never;
        };
        requestBody?: never;
        responses: {
            /** @description Null response */
            201: {
                headers: {
                    [name: string]: unknown;
                };
                content?: never;
            };
            /** @description unexpected error */
            default: {
                headers: {
                    [name: string]: unknown;
                };
                content: {
                    "application/json": components["schemas"]["Error"];
                };
            };
        };
    };
    showPetById: {
        parameters: {
            query?: never;
            header?: never;
            path: {
                /** @description The id of the pet to retrieve */
                petId: string;
            };
            cookie?: never;
        };
        requestBody?: never;
        responses: {
            /** @description Expected response to a valid request */
            200: {
                headers: {
                    [name: string]: unknown;
                };
                content: {
                    "application/json": components["schemas"]["Pet"];
                };
            };
            /** @description unexpected error */
            default: {
                headers: {
                    [name: string]: unknown;
                };
                content: {
                    "application/json": components["schemas"]["Error"];
                };
            };
        };
    };
}
export type Pet = components["schemas"]["Pet"];
export type Pets = components["schemas"]["Pets"];
export type Error = components["schemas"]["Error"];

export type Endpoints = keyof paths;
"
`;

exports[`e2e > petstore.json 1`] = `
"export interface paths {
    "/pets": {
        parameters: {
            query?: never;
            header?: never;
            path?: never;
            cookie?: never;
        };
        /** List all pets */
        get: operations["listPets"];
        put?: never;
        /** Create a pet */
        post: operations["createPets"];
        delete?: never;
        options?: never;
        head?: never;
        patch?: never;
        trace?: never;
    };
    "/pets/{petId}": {
        parameters: {
            query?: never;
            header?: never;
            path?: never;
            cookie?: never;
        };
        /** Info for a specific pet */
        get: operations["showPetById"];
        put?: never;
        post?: never;
        delete?: never;
        options?: never;
        head?: never;
        patch?: never;
        trace?: never;
    };
}
export type webhooks = Record<string, never>;
export interface components {
    schemas: {
        Pet: {
            /** Format: int64 */
            id: number;
            name: string;
            tag?: string;
        };
        Pets: components["schemas"]["Pet"][];
        Error: {
            /** Format: int32 */
            code: number;
            message: string;
        };
    };
    responses: never;
    parameters: never;
    requestBodies: never;
    headers: never;
    pathItems: never;
}
export type $defs = Record<string, never>;
export interface operations {
    listPets: {
        parameters: {
            query?: {
                /** @description How many items to return at one time (max 100) */
                limit?: number;
            };
            header?: never;
            path?: never;
            cookie?: never;
        };
        requestBody?: never;
        responses: {
            /** @description A paged array of pets */
            200: {
                headers: {
                    /** @description A link to the next page of responses */
                    "x-next"?: string;
                    [name: string]: unknown;
                };
                content: {
                    "application/json": components["schemas"]["Pets"];
                };
            };
            /** @description unexpected error */
            default: {
                headers: {
                    [name: string]: unknown;
                };
                content: {
                    "application/json": components["schemas"]["Error"];
                };
            };
        };
    };
    createPets: {
        parameters: {
            query?: never;
            header?: never;
            path?: never;
            cookie?: never;
        };
        requestBody?: never;
        responses: {
            /** @description Null response */
            201: {
                headers: {
                    [name: string]: unknown;
                };
                content?: never;
            };
            /** @description unexpected error */
            default: {
                headers: {
                    [name: string]: unknown;
                };
                content: {
                    "application/json": components["schemas"]["Error"];
                };
            };
        };
    };
    showPetById: {
        parameters: {
            query?: never;
            header?: never;
            path: {
                /** @description The id of the pet to retrieve */
                petId: string;
            };
            cookie?: never;
        };
        requestBody?: never;
        responses: {
            /** @description Expected response to a valid request */
            200: {
                headers: {
                    [name: string]: unknown;
                };
                content: {
                    "application/json": components["schemas"]["Pet"];
                };
            };
            /** @description unexpected error */
            default: {
                headers: {
                    [name: string]: unknown;
                };
                content: {
                    "application/json": components["schemas"]["Error"];
                };
            };
        };
    };
}
export type Pet = components["schemas"]["Pet"];
export type Pets = components["schemas"]["Pets"];
export type Error = components["schemas"]["Error"];

export type Endpoints = keyof paths;
"
`;
